File: sdks/go/cmd/elydora/plugins/codex_io.go
package plugins

import (
	"fmt"
	"os"
	"path/filepath"
)

func codexHomePath() (string, error) {
	configured := os.Getenv("CODEX_HOME")
	if configured == "" {
		home, err := os.UserHomeDir()
		if err != nil {
			return "", fmt.Errorf("resolve home directory: %w", err)
		}
		return filepath.Join(home, ".codex"), nil
	}
	info, err := os.Stat(configured)
	if err != nil {
		return "", fmt.Errorf("resolve CODEX_HOME at %s: %w", configured, err)
	}
	if !info.IsDir() {
		return "", fmt.Errorf("CODEX_HOME is not a directory: %s", configured)
	}
	canonical, err := filepath.EvalSymlinks(configured)
	if err != nil {
		return "", fmt.Errorf("canonicalize CODEX_HOME at %s: %w", configured, err)
	}
	canonical, err = filepath.Abs(canonical)
	if err != nil {
		return "", fmt.Errorf("resolve canonical CODEX_HOME at %s: %w", configured, err)
	}
	exists, err := managedPhysicalDirectoryExists(canonical, "CODEX_HOME")
	if err != nil {
		return "", err
	}
	if !exists {
		return "", fmt.Errorf("CODEX_HOME is missing: %s", canonical)
	}
	return canonical, nil
}

func codexConfigPath() (string, error) {
	home, err := codexHomePath()
	if err != nil {
		return "", err
	}
	return filepath.Join(home, codexConfigFile), nil
}

func readCodexDocument() (*codexDocument, error) {
	filePath, err := codexConfigPath()
	if err != nil {
		return nil, err
	}
	directory := filepath.Dir(filePath)
	if _, err := managedPhysicalDirectoryExists(directory, "Codex hooks directory"); err != nil {
		return nil, err
	}
	snapshot, err := readManagedFile(filePath, "Codex user hooks", maxManagedSourceBytes)
	if err != nil {
		return nil, err
	}
	if snapshot == nil {
		return createCodexDocument(filePath), nil
	}
	return parseCodexDocument(filePath, snapshot.contents)
}

func prepareRenderedCodexChange(rendered *codexRenderedDocument) (*fileChange, error) {
	if rendered == nil || !rendered.changed {
		return nil, nil
	}
	return prepareSourceChange(
		rendered.document.filePath,
		"Codex user hooks",
		rendered.document.raw,
		rendered.document.exists,
		rendered.next,
		0600,
		rendered.remove,
	)
}

func codexRuntimeFilesExist(contracts []managedRuntimeContract) (bool, error) {
	return managedRuntimeFilesPresent(contracts, codexAgentKey)
}

File: sdks/node/src/plugins/codex-io.ts
import fsp from 'node:fs/promises';
import os from 'node:os';
import path from 'node:path';
import {
  AGENT_KEY,
  CONFIG_FILE,
  type CodexDocument,
  type RenderedDocument,
  type RuntimeContract,
  createDocument,
  parseDocument,
} from './codex-contract.js';
import { asError, errorMessage } from './common.js';
import { inspectPhysicalDirectory, readPhysicalFile } from './managed-files.js';
import { managedRuntimePresent } from './managed-runtime-status.js';
import {
  commitManagedTransaction,
  prepareManagedFileChange,
} from './managed-transaction.js';

async function codexHomePath(): Promise<string> {
  const configured = process.env.CODEX_HOME;
  if (configured === undefined || configured === '') return path.join(os.homedir(), '.codex');
  let metadata;
  try {
    metadata = await fsp.stat(configured);
  } catch (error) {
    throw new Error(`Resolve CODEX_HOME at ${configured}: ${errorMessage(error)}`, {
      cause: asError(error),
    });
  }
  if (!metadata.isDirectory()) throw new Error(`CODEX_HOME is not a directory: ${configured}`);
  let canonical;
  try {
    canonical = await fsp.realpath(configured);
  } catch (error) {
    throw new Error(`Canonicalize CODEX_HOME at ${configured}: ${errorMessage(error)}`, {
      cause: asError(error),
    });
  }
  if (!await inspectPhysicalDirectory(canonical, 'CODEX_HOME')) {
    throw new Error(`CODEX_HOME is missing: ${canonical}`);
  }
  return canonical;
}

export async function codexConfigPath(): Promise<string> {
  return path.join(await codexHomePath(), CONFIG_FILE);
}

export async function readDocument(): Promise<CodexDocument> {
  const filePath = await codexConfigPath();
  const snapshot = await readPhysicalFile(filePath, 'Codex user hooks');
  return snapshot ? parseDocument(filePath, snapshot.contents) : createDocument(filePath);
}

export async function writeDocument(rendered: RenderedDocument): Promise<void> {
  if (!rendered.changed) return;
  const change = await prepareManagedFileChange({
    filePath: rendered.document.filePath,
    label: 'Codex user hooks',
    next: rendered.next,
    mode: 0o600,
    expectedSource: rendered.document.raw,
    verifyExpectedSource: true,
  });
  if (!change) return;
  await commitManagedTransaction({
    displayName: 'Codex',
    operation: 'uninstall',
    directories: [{
      path: path.dirname(rendered.document.filePath),
      label: 'Codex hooks directory',
    }],
    changes: [change],
  });
}

export async function runtimeFilesExist(contracts: RuntimeContract[]): Promise<boolean> {
  for (const contract of contracts) {
    if (await managedRuntimePresent(contract, AGENT_KEY)) return true;
  }
  return false;
}
